pkg/generic/thrift/write.go

```diff
@@ -690,36 +682,51 @@ func writeStruct(ctx context.Context, val interface{}, out thrift.TProtocol, t *
 			}
 			continue
 		}
-		if !ok || elem == nil {
+
+		if field.ValueMapping != nil {
+			elem, err = field.ValueMapping.Request(ctx, elem, field)
+			if err != nil {
+				return err
+			}
+		}
+
+		// empty fields
+		if elem == nil || !ok {
 			if !field.Optional {
-				elem, _, err = getDefaultValueAndWriter(field.Type, opt)
-				if err != nil {
+				// empty fields don't need value-mapping here, since writeEmptyValue decides zero value based on Thrift type
+				if err := out.WriteFieldBegin(field.Name, field.Type.Type.ToThriftTType(), int16(field.ID)); err != nil {
+					return err
+				}
+				if err := writeEmptyValue(out, field.Type, opt); err != nil {
 					return fmt.Errorf("field (%d/%s) error: %w", field.ID, name, err)
 				}
-			} else {
-				continue
+				if err := out.WriteFieldEnd(); err != nil {
+					return err
+				}
 			}
-		}
-		if field.ValueMapping != nil {
-			elem, err = field.ValueMapping.Request(ctx, elem, field)
+		} else { // normal fields
+			if field.ValueMapping != nil {
```

Comment:
If you do field.ValueMapping above on L686, is this necessary to do?

Reply:
forget to delete, I will fix it